WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace WalletWasabi.Backend.Filters
+{
+	public class IdempotentActionFilterAttributeImpl : ActionFilterAttribute
+	{
+		private readonly IMemoryCache _cache;
+
+		public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
+			: base()
+		{
+			_cache = cache;
+		}
+
+		public override void OnActionExecuting(ActionExecutingContext context)
+		{
+			var request = context.HttpContext.Request;
+
+			var (cacheKey, body) = GetCacheEntryKey(request);
+
+			if (_cache.TryGetValue<OkObjectResult>(cacheKey, out var cachedResponse))
+			{
+				context.Result = cachedResponse;
+				return;
+			}
+
+			context.HttpContext.Items["cached-key"] = cacheKey;
+		}
+
+		public override void OnResultExecuted(ResultExecutedContext context)
+		{
+			var cacheKey = context.HttpContext.Items["cached-key"];
+
+			_cache.Set(cacheKey, context.Result, DateTimeOffset.UtcNow.AddMinutes(5));
+		}
```

Comment:
I think this should be conditional on `context.Exception` and also that `context.Result` is successful (at the protocol level as well as http level), otherwise anyone can fill the cache with error responses to invalid requests

Reply:
`OnResultExecuted` is invoked only after a successful response. It is not possible to fill the cache with error response.